`lsp-restart` cannot restart lsp correctly.

### Summary

When I restart lsp in helix, lsp will not work prefectly again.

### Reproduction Steps

<!-- Ideally provide a key sequence and/or asciinema.org recording. --> 

I tried this:

1. run master
```bash
cd helix
cargo r -- -vvv
```
2. open readme.md and check lsp is ok.
3. `:lsp-restart`
4. open `main.rs` and check lsp is ok.
5. `:lsp-restart` and check
6. open log

I expected this to happen:

Lsp work again.

Instead, this happened:

marksman get error.
rust-analyzer get error, like `completion` is broken.
gopls is broken.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```

2024-04-27T10:21:47.182 helix_lsp::transport [INFO] marksman <- {"jsonrpc":"2.0","id":2,"result":null}
2024-04-27T10:21:47.182 helix_lsp::transport [INFO] marksman <- null
2024-04-27T10:21:47.182 helix_lsp::transport [INFO] marksman -> {"jsonrpc":"2.0","method":"exit"}
2024-04-27T10:21:47.186 helix_view::document [DEBUG] id 2 modified - last saved: 0, current: 0
- 2024-04-27T10:21:47.196 helix_lsp::transport [ERROR] marksman err: <- StreamClosed
2024-04-27T10:21:47.196 mio::poll [TRACE] deregistering event source from poller
2024-04-27T10:21:47.196 mio::poll [TRACE] deregistering event source from poller
2024-04-27T10:21:47.196 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v1), Notification(Notification { jsonrpc: None, method: "exit", params: None })))
2024-04-27T10:21:47.196 helix_view::editor [DEBUG] editor status: Language server exited
2024-04-27T10:21:47.196 mio::poll [TRACE] deregistering event source from poller
2024-04-27T10:21:47.196 helix_lsp::file_event [DEBUG] Removing LSP client: 1v1
2024-04-27T10:21:47.196 mio::poll [TRACE] deregistering event source from poller
2024-04-27T10:21:47.230 helix_term::application [DEBUG] received editor event: Redraw
2024-04-27T10:21:47.237 helix_view::document [DEBUG] id 2 modified - last saved: 0, current: 0
2024-04-27T10:21:47.562 helix_lsp::transport [ERROR] marksman err <- "[10:21:47 INF] <LSP Entry> Starting Marksman LSP server: {}\n"
2024-04-27T10:21:48.679 helix_lsp::transport [INFO] marksman <- {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":{"openClose":true,"change":1},"hoverProvider":true,"completionProvider":{"triggerCharacters":["[","#","("]},"definitionProvider":true,"referencesProvider":true,"documentSymbolProvider":true,"workspaceSymbolProvider":true,"codeActionProvider":{"resolveProvider":false},"codeLensProvider":{},"renameProvider":{"prepareProvider":true},"executeCommandProvider":{"commands":[]},"semanticTokensProvider":{"legend":{"tokenTypes":["class","class","enumMember"],"tokenModifiers":[]},"range":true,"full":{"delta":false}},"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true},"fileOperations":{"didCreate":{"filters":[{"pattern":{"glob":"**/*.{md,markdown}","matches":"file","options":{"ignoreCase":true}}}]},"didDelete":{"filters":[{"pattern":{"glob":"**/*.{md,markdown}","matches":"file","options":{"ignoreCase":true}}}]}}}}}}
2024-04-27T10:21:48.679 helix_lsp::transport [INFO] marksman <- {"capabilities":{"codeActionProvider":{"resolveProvider":false},"codeLensProvider":{},"completionProvider":{"triggerCharacters":["[","#","("]},"definitionProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":[]},"hoverProvider":true,"referencesProvider":true,"renameProvider":{"prepareProvider":true},"semanticTokensProvider":{"full":{"delta":false},"legend":{"tokenModifiers":[],"tokenTypes":["class","class","enumMember"]},"range":true},"textDocumentSync":{"change":1,"openClose":true},"workspace":{"fileOperations":{"didCreate":{"filters":[{"pattern":{"glob":"**/*.{md,markdown}","matches":"file","options":{"ignoreCase":true}}}]},"didDelete":{"filters":[{"pattern":{"glob":"**/*.{md,markdown}","matches":"file","options":{"ignoreCase":true}}}]}},"workspaceFolders":{"changeNotifications":true,"supported":true}},"workspaceSymbolProvider":true}}
2024-04-27T10:21:48.679 helix_lsp::transport [INFO] marksman -> {"jsonrpc":"2.0","method":"initialized","params":{}}
2024-04-27T10:21:48.679 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(2v1), Notification(Notification { jsonrpc: None, method: "initialized", params: None })))
2024-04-27T10:21:48.679 helix_term::application [WARN] can't find language server with id `2v1`
2024-04-27T10:21:48.714 helix_term::application [DEBUG] received editor event: Redraw
2024-04-27T10:21:48.722 helix_view::document [DEBUG] id 2 modified - last saved: 0, current: 0
2024-04-27T10:21:48.967 helix_lsp::transport [INFO] marksman <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/erasin/Development/helix/README.md","diagnostics":[{"range":{"start":{"line":53,"character":37},"end":{"line":53,"character":67}},"severity":2,"code":"1","source":"Marksman","message":"Ambiguous link to document './docs/CONTRIBUTING.md'","relatedInformation":[{"location":{"uri":"file:///home/erasin/Development/helix/docs/CONTRIBUTING.md","range":{"start":{"line":21,"character":0},"end":{"line":21,"character":30}}},"message":"Duplicate definition of document './docs/CONTRIBUTING.md'"},{"location":{"uri":"file:///home/erasin/Development/helix/docs/CONTRIBUTING.md","range":{"start":{"line":0,"character":0},"end":{"line":0,"character":14}}},"message":"Duplicate definition of document './docs/CONTRIBUTING.md'"},{"location":{"uri":"file:///home/erasin/Development/helix/docs/CONTRIBUTING.md","range":{"start":{"line":41,"character":0},"end":{"line":41,"character":9}}},"message":"Duplicate definition of document './docs/CONTRIBUTING.md'"}]}]}}

2024-04-27T10:21:56.298 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","id":0,"method":"workspace/configuration","params":{"items":[{"section":"rust-analyzer"}]}}
2024-04-27T10:21:56.298 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","id":1,"method":"window/workDoneProgress/create","params":{"token":"rustAnalyzer/Fetching"}}
2024-04-27T10:21:56.298 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v3), MethodCall(MethodCall { jsonrpc: Some(V2), method: "workspace/configuration", params: Map({"items": Array [Object {"section": String("rust-analyzer")}]}), id: Num(0) })))
2024-04-27T10:21:56.298 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Fetching","value":{"kind":"begin","title":"Fetching","cancellable":false}}}
2024-04-27T10:21:56.298 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v3), MethodCall(MethodCall { jsonrpc: Some(V2), method: "window/workDoneProgress/create", params: Map({"token": String("rustAnalyzer/Fetching")}), id: Num(1) })))
2024-04-27T10:21:56.298 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","result":[null],"id":0}
2024-04-27T10:21:56.298 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","result":null,"id":1}
2024-04-27T10:21:56.298 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v3), Notification(Notification { jsonrpc: Some(V2), method: "$/progress", params: Map({"token": String("rustAnalyzer/Fetching"), "value": Object {"cancellable": Bool(false), "kind": String("begin"), "title": String("Fetching")}}) })))
2024-04-27T10:21:56.330 helix_term::application [DEBUG] received editor event: Redraw
2024-04-27T10:21:56.338 helix_view::document [DEBUG] id 3 modified - last saved: 0, current: 0
2024-04-27T10:21:56.339 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Fetching","value":{"kind":"report","cancellable":false,"message":"metadata"}}}
2024-04-27T10:21:56.340 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v3), Notification(Notification { jsonrpc: Some(V2), method: "$/progress", params: Map({"token": String("rustAnalyzer/Fetching"), "value": Object {"cancellable": Bool(false), "kind": String("report"), "message": String("metadata")}}) })))
2024-04-27T10:21:56.374 helix_term::application [DEBUG] received editor event: Redraw
2024-04-27T10:21:56.382 helix_view::document [DEBUG] id 3 modified - last saved: 0, current: 0
2024-04-27T10:21:56.490 helix_term::application [DEBUG] received editor event: IdleTimer
2024-04-27T10:21:56.513 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","id":2,"method":"client/registerCapability","params":{"registrations":[{"id":"workspace/didChangeWatchedFiles","method":"workspace/didChangeWatchedFiles","registerOptions":{"watchers":[{"globPattern":"/home/erasin/Development/helix/helix-core/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-core/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-core/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-dap/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-dap/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-dap/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-event/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-event/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-event/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-loader/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-loader/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-loader/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-lsp/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-lsp/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-lsp/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-parsec/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-parsec/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-parsec/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-stdx/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-stdx/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-stdx/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-term/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-term/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-term/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-tui/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-tui/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-tui/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-vcs/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-vcs/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-vcs/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/helix-view/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/helix-view/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/helix-view/**/Cargo.lock"},{"globPattern":"/home/erasin/Development/helix/xtask/**/*.rs"},{"globPattern":"/home/erasin/Development/helix/xtask/**/Cargo.toml"},{"globPattern":"/home/erasin/Development/helix/xtask/**/Cargo.lock"}]}}]}}
2024-04-27T10:21:56.514 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v3), MethodCall(MethodCall { jsonrpc: Some(V2), method: "client/registerCapability", params: Map({"registrations": Array [Object {"id": String("workspace/didChangeWatchedFiles"), "method": String("workspace/didChangeWatchedFiles"), "registerOptions": Object {"watchers": Array [Object {"globPattern": String("/home/erasin/Development/helix/helix-core/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-core/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-core/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-dap/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-dap/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-dap/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-event/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-event/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-event/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-loader/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-loader/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-loader/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-lsp/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-lsp/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-lsp/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-parsec/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-parsec/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-parsec/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-stdx/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-stdx/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-stdx/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-term/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-term/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-term/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-tui/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-tui/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-tui/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-vcs/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-vcs/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-vcs/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-view/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-view/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/helix-view/**/Cargo.lock")}, Object {"globPattern": String("/home/erasin/Development/helix/xtask/**/*.rs")}, Object {"globPattern": String("/home/erasin/Development/helix/xtask/**/Cargo.toml")}, Object {"globPattern": String("/home/erasin/Development/helix/xtask/**/Cargo.lock")}]}}]}), id: Num(2) })))
2024-04-27T10:21:56.514 helix_lsp::file_event [DEBUG] Registering didChangeWatchedFiles for client '1v3' with id 'workspace/didChangeWatchedFiles'
2024-04-27T10:21:56.514 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","result":null,"id":2}
2024-04-27T10:21:56.515 globset [DEBUG] built glob set; 0 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 3 required extensions, 0 regexes
2024-04-27T10:21:56.528 helix_lsp::transport [ERROR] rust-analyzer err <- "[ERROR project_model::workspace] cyclic deps: helix_view(Idx::<CrateData>(33)) -> helix_tui(Idx::<CrateData>(26)), alternative path: helix_tui(Idx::<CrateData>(26)) -> helix_view(Idx::<CrateData>(33))\n"
2024-04-27T10:21:56.532 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Fetching","value":{"kind":"end"}}}
2024-04-27T10:21:56.533 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","id":3,"method":"window/workDoneProgress/create","params":{"token":"rustAnalyzer/Roots Scanned"}}
2024-04-27T10:21:56.533 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"begin","title":"Roots Scanned","cancellable":false,"message":"0/215","percentage":0}}}
2024-04-27T10:21:56.533 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"report","cancellable":false,"message":"0/215: /home/erasin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/adler-1.0.2","percentage":0}}}
2024-04-27T10:21:56.533 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v3), Notification(Notification { jsonrpc: Some(V2), method: "$/progress", params: Map({"token": String("rustAnalyzer/Fetching"), "value": Object {"kind": String("end")}}) })))
2024-04-27T10:21:56.533 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"report","cancellable":false,"message":"0/215: /home/erasin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/adler-1.0.2","percentage":0}}}
2024-04-27T10:21:56.533 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"report","cancellable":false,"message":"0/215: /home/erasin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/adler-1.0.2/src","percentage":0}}}
2024-04-27T10:21:56.533 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"report","cancellable":false,"message":"0/215: /home/erasin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/adler-1.0.2/.github","percentage":0}}}
2024-04-27T10:21:56.533 helix_term::application [DEBUG] received editor event: LanguageServerMessage((LanguageServerId(1v3), MethodCall(MethodCall { jsonrpc: Some(V2), method: "window/workDoneProgress/create", params: Map({"token": String("rustAnalyzer/Roots Scanned")}), id: Num(3) })))
2024-04-27T10:22:24.086 helix_lsp::transport [ERROR] rust-analyzer err <- "[ERROR rust_analyzer::handlers::notification] unexpected DidChangeTextDocument: /home/erasin/Development/helix/helix-term/src/main.rs\n"
2024-04-27T10:22:24.093 helix_view::document [DEBUG] id 4 modified - last saved: 0, current: 0
2024-04-27T10:22:24.332 helix_term::application [DEBUG] received editor event: IdleTimer
2024-04-27T10:22:24.337 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","method":"textDocument/completion","params":{"context":{"triggerKind":1},"position":{"character":15,"line":101},"textDocument":{"uri":"file:///home/erasin/Development/helix/helix-term/src/main.rs"}},"id":3}
2024-04-27T10:22:24.344 helix_view::document [DEBUG] id 4 modified - last saved: 0, current: 0
2024-04-27T10:22:25.286 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":15,"line":101},"start":{"character":14,"line":101}},"text":""}],"textDocument":{"uri":"file:///home/erasin/Development/helix/helix-term/src/main.rs","version":5}}}
2024-04-27T10:22:25.286 helix_lsp::transport [ERROR] rust-analyzer err <- "[ERROR rust_analyzer::handlers::notification] unexpected DidChangeTextDocument: /home/erasin/Development/helix/helix-term/src/main.rs\n"
2024-04-27T10:22:25.293 helix_view::document [DEBUG] id 4 modified - last saved: 0, current: 0
2024-04-27T10:22:25.498 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":14,"line":101},"start":{"character":13,"line":101}},"text":""}],"textDocument":{"uri":"file:///home/erasin/Development/helix/helix-term/src/main.rs","version":6}}}
2024-04-27T10:22:25.498 helix_lsp::transport [ERROR] rust-analyzer err <- "[ERROR rust_analyzer::handlers::notification] unexpected DidChangeTextDocument: /home/erasin/Development/helix/helix-term/src/main.rs\n"
2024-04-27T10:22:55.093 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":9,"line":93},"start":{"character":9,"line":93}},"text":"d"}],"textDocument":{"uri":"file:///home/erasin/Development/helix/helix-term/src/main.rs","version":52}}}
2024-04-27T10:22:55.094 helix_lsp::transport [ERROR] rust-analyzer err <- "[ERROR rust_analyzer::handlers::notification] unexpected DidChangeTextDocument: /home/erasin/Development/helix/helix-term/src/main.rs\n"
2024-04-27T10:22:55.103 helix_view::document [DEBUG] id 4 modified - last saved: 0, current: 7
2024-04-27T10:22:55.285 helix_core::auto_pairs [TRACE] autopairs hook selection: Selection {
    ranges: [
        Range {
            anchor: 3359,
            head: 3360,
            old_visual_position: None,
        },
    ],
    primary_index: 0,
}
2024-04-27T10:22:55.289 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":10,"line":93},"start":{"character":10,"line":93}},"text":"i"}],"textDocument":{"uri":"file:///home/erasin/Development/helix/helix-term/src/main.rs","version":53}}}
2024-04-27T10:22:55.289 helix_lsp::transport [ERROR] rust-analyzer err <- "[ERROR rust_analyzer::handlers::notification] unexpected DidChangeTextDocument: /home/erasin/Development/helix/helix-term/src/main.rs\n"
2024-04-27T10:22:55.299 helix_view::document [DEBUG] id 4 modified - last saved: 0, current: 7
2024-04-27T10:22:55.537 helix_term::application [DEBUG] received editor event: IdleTimer
2024-04-27T10:22:55.838 helix_view::document [DEBUG] id 4 modified - last saved: 0, current: 8
2024-04-27T10:22:56.077 helix_term::application [DEBUG] received editor event: IdleTimer
```

</details>


### Platform

linux

### Terminal Emulator

alacritty release

### Installation Method

source

### Helix Version

helix (5ee7411)

--- a/helix-lsp/src/lib.rs
+++ b/helix-lsp/src/lib.rs
@@ -736,35 +736,30 @@ impl Registry {
             .language_servers
             .iter()
             .filter_map(|LanguageServerFeatures { name, .. }| {
-                if self.inner_by_name.contains_key(name) {
-                    let client = match self.start_client(
-                        name.clone(),
-                        language_config,
-                        doc_path,
-                        root_dirs,
-                        enable_snippets,
-                    ) {
-                        Ok(client) => client,
-                        Err(StartupError::NoRequiredRootFound) => return None,
-                        Err(StartupError::Error(err)) => return Some(Err(err)),
-                    };
-                    let old_clients = self
-                        .inner_by_name
-                        .insert(name.clone(), vec![client.clone()])
-                        .unwrap();
-
+                if let Some(old_clients) = self.inner_by_name.remove(name) {
                     for old_client in old_clients {
                         self.file_event_handler.remove_client(old_client.id());
-                        self.inner.remove(client.id());
+                        self.inner.remove(old_client.id());
                         tokio::spawn(async move {
                             let _ = old_client.force_shutdown().await;
                         });
                     }
-
-                    Some(Ok(client))
-                } else {
-                    None
                 }
+                let client = match self.start_client(
+                    name.clone(),
+                    language_config,
+                    doc_path,
+                    root_dirs,
+                    enable_snippets,
+                ) {
+                    Ok(client) => client,
+                    Err(StartupError::NoRequiredRootFound) => return None,
+                    Err(StartupError::Error(err)) => return Some(Err(err)),
+                };
+                self.inner_by_name
+                    .insert(name.to_owned(), vec![client.clone()]);
+
+                Some(Ok(client))
             })
             .collect()
     }

